nf1_cellprofiler_classifier_fs_all_cellprofiler_all_feat_perfect.py

```diff
@@ -0,0 +1,250 @@
+
+featdf = featdf[feat_col]
+featdf = featdf.loc[:,featdf.columns != 'Unnamed: 0'] # Remove the unamed column
+
+
+# # Sampling
+
+# In[8]:
+
+
+min_class_samps_size = min(featdf['genotype_label'].value_counts().values) # Sample size for the minority class
+dsamp_featdf = featdf.groupby('genotype_label', group_keys=False).apply(lambda x: x.sample(n=min_class_samps_size, random_state=rnd_val)) # Down-sample features according to minority class
+traindf, testdf = train_test_split(dsamp_featdf, random_state=rnd_val, shuffle=True, train_size=0.9) # Train and Test Dataframes
+
+
+# In[9]:
+
+
+train_feat = traindf.to_numpy()
+test_feat = testdf.to_numpy()
+
+
+# # Principle Component Analysis (PCA) Visualization
+
+# In[10]:
+
+
+null_idx = np.nonzero(train_feat[:,-1] == 0)[0]
+wt_idx = np.nonzero(train_feat[:,-1] != 0)[0]
+
+
+# In[11]:
+
+
+pca = PCA(n_components=2)
+pca_features = pca.fit_transform(train_feat[:,0:-1])
+print(f'Explained variance in PC1 and PC2 = {np.sum(pca.explained_variance_ratio_)}')
+null = plt.scatter(pca_features[null_idx,0],pca_features[null_idx,1], marker='x', color='r')
+wt = plt.scatter(pca_features[wt_idx,0],pca_features[wt_idx,1], marker='.', color='b')
+plt.xlabel('PC1')
+plt.ylabel('PC2')
+plt.legend((null, wt), ('null', 'wt'))
+plt.show()
+
+
+# In[12]:
+
+
+reducer = umap.UMAP(random_state=rnd_val)
+reducer.fit(train_feat[:,0:-1])
+
+
+# In[28]:
+
+
+null = plt.scatter(reducer.embedding_[null_idx,0],reducer.embedding_[null_idx,1], marker='x', color='r')
+wt = plt.scatter(reducer.embedding_[wt_idx,0],reducer.embedding_[wt_idx,1], marker='.', color='b')
+plt.title('Embedding of the training set by UMAP', fontsize=24)
+plt.legend((null, wt), ('null', 'wt'), loc='lower right')
+plt.show()
+
+
+# # K Cross Validation
+
+# In[14]:
+
+
+num_splits = 5 # Default number of splits
+
+def kcross_val(model, feat, splits=num_splits):
+
+    cv = KFold(n_splits = splits, random_state=rnd_val, shuffle=True)
+    max_acc = 0 # Keep track of the maximum accuracy
+
+    for idx, (train_idx, val_idx) in enumerate(cv.split(X=feat)):
+        yval = feat[val_idx,-1].astype(np.int64)
+        ytrain = feat[train_idx,-1].astype(np.int64)
+        Xval = feat[val_idx,0:-1]
+        Xtrain = feat[train_idx,0:-1]
+
+        model.fit(Xtrain, ytrain)
+        acc = model.score(Xval, yval)
+
+        if max_acc < acc: # Find the model with the best validation accuracy
+            
+            max_acc, best_model, best_preds, val_labels = acc, model, model.predict(Xval), yval
+    
+    res = {'acc':max_acc, 'model':best_model, 'preds':best_preds, 'labels':val_labels} # Store as a dictionary
+
+    return res
+
+
+# In[15]:
+
+
+# Returns the naive accuracy:
+def naive_acc(labels):
+    naive_pred = rng.integers(low=0, high=2, size=labels.shape[0])
+    naive_acc = accuracy_score(naive_pred,labels)
+    return naive_acc
```

Comment:
Consider using another name besides that of the function name for this variable. Doing this will increase readability by distinguishing between the function and the value here.

Reply:
I made it a small class (similar to sklearn). Let me know if you think I should do something else